File: src/Accountant.Persistence/Repositories/SyncRepository.cs
using Accountant.Application.Contracts.Sync;
using Domain.Accountant;

namespace Persistence.Repositories.Accountant;

public class SyncRepository : BaseRepository, ISyncRepository
{
    public SyncRepository(PersonalAssistantContext efContext)
        : base(efContext) { }

    public async Task SyncAsync(IEnumerable<Account> accounts, IEnumerable<Category> categories, IEnumerable<Transaction> transactions,
        IEnumerable<UpcomingExpense> upcomingExpenses, IEnumerable<Debt> debts, IEnumerable<AutomaticTransaction> automaticTransactions)
    {
        EFContext.Accounts.AddRange(accounts);
        EFContext.Categories.AddRange(categories);

        List<Transaction> transactionsList = transactions.ToList();
        transactionsList.ForEach(x => x.Date = x.Date.ToUniversalTime());
        EFContext.Transactions.AddRange(transactionsList);

        List<UpcomingExpense> upcomingExpensesList = upcomingExpenses.ToList();
        upcomingExpensesList.ForEach(x => x.Date = x.Date.ToUniversalTime());
        EFContext.UpcomingExpenses.AddRange(upcomingExpensesList);

        EFContext.Debts.AddRange(debts);
        EFContext.AutomaticTransactions.AddRange(automaticTransactions);

        await EFContext.SaveChangesAsync();
    }
}

File: src/Infrastructure/Accountant.Persistence.Fs/SyncRepository.fs
namespace Accountant.Persistence.Fs

open Npgsql
open ConnectionUtils
open Models
open Sentry
type Transaction = Accountant.Persistence.Fs.Models.Transaction

module SyncRepository =

    let sync
        (accounts: Account list)
        (categories: Category list)
        (transactions: Transaction list)
        (upcomingExpenses: UpcomingExpense list)
        (debts: Debt list)
        (automaticTransactions: AutomaticTransaction list)
        (userId: int)
        connectionString
        (metricsSpan: ISpan)
        =
        let metric = metricsSpan.StartChild("SyncRepository.sync")

        use conn = new NpgsqlConnection(connectionString)
        conn.Open()

        use tr = conn.BeginTransaction()

        let accountIds =
            accounts
            |> List.map (fun account -> { account with UserId = userId } )
            |> List.map (fun account ->
                AccountsRepository.create account (TransactionConnection(conn)) metric |> Async.AwaitTask
            )
            |> Async.Sequential
            |> Async.RunSynchronously

        let categoryIds =
            categories
            |> List.map (fun account -> { account with UserId = userId } )
            |> List.map (fun category ->
                CategoriesRepository.create category (TransactionConnection(conn)) metric |> Async.AwaitTask
            )
            |> Async.Sequential
            |> Async.RunSynchronously

        let transactionIds =
            transactions
            |> List.map (fun transaction ->
                TransactionsRepository.create transaction (TransactionConnection(conn)) (Some(tr)) metric |> Async.AwaitTask
            )
            |> Async.Sequential
            |> Async.RunSynchronously

        let upcomingExpenseIds =
            upcomingExpenses
            |> List.map (fun account -> { account with UserId = userId } )
            |> List.map (fun upcomingExpense ->
                UpcomingExpensesRepository.create upcomingExpense (TransactionConnection(conn)) metric |> Async.AwaitTask
            )
            |> Async.Sequential
            |> Async.RunSynchronously

        let debtIds =
            debts
            |> List.map (fun account -> { account with UserId = userId } )
            |> List.map (fun debt ->
                    DebtsRepository.create debt (TransactionConnection(conn)) metric |> Async.AwaitTask
            )
            |> Async.Sequential
            |> Async.RunSynchronously

        let automaticTransactionIds =
            automaticTransactions
            |> List.map (fun account -> { account with UserId = userId } )
            |> List.map (fun automaticTransaction ->
                    AutomaticTransactionsRepository.create automaticTransaction (TransactionConnection(conn)) metric |> Async.AwaitTask
            )
            |> Async.Sequential
            |> Async.RunSynchronously

        tr.Commit()

        metric.Finish()

        (accountIds, categoryIds, transactionIds, upcomingExpenseIds, debtIds, automaticTransactionIds)
